File: src/main/kotlin/net/merayen/elastic/backend/architectures/local/nodes/signalgenerator_1/LProcessor.java
package net.merayen.elastic.backend.architectures.local.nodes.signalgenerator_1;

import net.merayen.elastic.backend.architectures.local.LocalProcessor;
import net.merayen.elastic.backend.architectures.local.lets.*;
import net.merayen.elastic.backend.architectures.local.nodes.poly_1.SessionKeeper;
import net.merayen.elastic.backend.architectures.local.utils.InputSignalParametersProcessor;
import net.merayen.elastic.backend.midi.MidiState;
import net.merayen.elastic.backend.util.AudioUtil;

import java.util.*;

public class LProcessor extends LocalProcessor implements SessionKeeper {
	private enum Mode {
		NOTHING,
		RAW,
		MIDI,
		FREQUENCY,
		MALFUNCTION // E.g, wrong line has been connected. We output only zeroes in this case (and should send a warning)
	}

	private LNode lnode;
	private Resampling resampling;

	private Mode mode;

	private float[][] input_frequency_buffer;

	private List<short[]> keys_down = new ArrayList<>();

	private boolean sustain;
	private float volume;
	private short[] active_tangent;

	private double pos = new Random().nextDouble() * Math.PI * 2;

	private MidiState midiState = new MidiState() {
		@Override
		protected void onPitchBendSensitivityChange(float v) {}

		@Override
		protected void onKeyDown(short tangent, float _velocity) {
			if (!sustain)
				keys_down.add(getCurrentMidiPacket());

			active_tangent = getCurrentMidiPacket();
		}

		@Override
		protected void onKeyUp(short tangent) {
			keys_down.removeIf(m -> m[1] == tangent);
			if (!sustain)
				active_tangent = keys_down.isEmpty() ? null : keys_down.get(keys_down.size() - 1);
		}

		@Override
		protected void onSustain(float value) {
			/*if (sustain && getCurrentMidiPacket()[2] == 0) {
				if (keys_down.isEmpty())
					active_tangent = null;
				else
					active_tangent = keys_down.get(keys_down.size() - 1);
			}

			sustain = getCurrentMidiPacket()[2] != 0;*/
		}

		@Override
		protected void onVolumeChange(float volume) {}

		@Override
		protected void onPitchChange(float semitones) {
			//System.out.println(getSessionID() + "\t" + semitones);
		}
	};

	@Override
	protected void onInit() {
		lnode = (LNode)getLocalNode();

		Inlet frequency = getInlet("frequency");
		Outlet output = getOutlet("output");

		if(output == null)
			mode = Mode.NOTHING;
		if(frequency == null) {
			mode = Mode.RAW;
		} else {
			if(frequency instanceof AudioInlet) {
				mode = Mode.FREQUENCY;
			} else if(frequency instanceof MidiInlet) {
				mode = Mode.MIDI;
			} else {
				mode = Mode.MALFUNCTION; // We don't understand the input on the frequency-port
			}
		}
	}

	@Override
	public void onProcess() {
		if (frameFinished())
			return;

		if(mode == Mode.RAW)
			generateRaw();
		else if(mode == Mode.FREQUENCY)
			generateWithFrequency();
		else if(mode == Mode.MIDI)
			generateWithMidi();
		else
			throw new RuntimeException("Not implemented");
	}

	private void generateRaw() {
		AudioOutlet outlet = (AudioOutlet)getOutlet("output");

		if(resampling != null && !outlet.available())
			resampling.update(lnode.frequency);

		outlet.push();
	}

	private void generateWithFrequency() {
		if (!available())
			return;

		AudioOutlet outlet = (AudioOutlet) getOutlet("output");
		AudioInlet frequency = (AudioInlet) getInlet("frequency");

		// Transform input according to the UI's InputSignalProcessor() parameters
		if (input_frequency_buffer == null)
			input_frequency_buffer = new float[1][buffer_size];

		InputSignalParametersProcessor.process(lnode, "frequency", new float[][]{frequency.outlet.audio[0]}, input_frequency_buffer, 0, buffer_size);

		int i;
		for (i = 0; i < buffer_size; i++) {
			outlet.audio[0][i] = lnode.curve_wave[Math.floorMod((int) (pos / (Math.PI * 2 * sampleRate) * lnode.curve_wave.length), lnode.curve_wave.length)];
			pos += input_frequency_buffer[0][i];
		}

		outlet.push();
	}

	private void generateWithMidi() {
		if(!available())
			return;

		AudioOutlet outlet = (AudioOutlet)getOutlet("output");
		MidiInlet inlet = (MidiInlet)getInlet("frequency");

		updateKeysDown();

		float pitch = midiState.getPitch();

		if(active_tangent != null && resampling != null) {
			double freq = AudioUtil.midiNoteToFreq(active_tangent[1] + pitch);
			float volume_div = sampleRate / 1000f;
			float[] audio = outlet.audio[0];

			resampling.update((float)freq, buffer_size);
			float new_volume = (float)Math.pow(midiState.getVolume(), 3) * midiState.getVelocity();

			if (volume != new_volume) {
				for (int i = 0; i < buffer_size; i++) {
					volume += (new_volume - volume) / volume_div;
					audio[i] *= volume;
				}
			} else if(volume != 1f) {
				for (int i = 0; i < buffer_size; i++)
					audio[i] *= volume;
			}

		} else { // No key down? Silence!
			pos = new Random().nextDouble() * Math.PI * 2;
			for(int i = 0; i < buffer_size; i++)
				outlet.audio[0][i] = 0;
		}

//		System.out.printf("signalgenerator_1 pushing audio out for session %s (%f)\n", getSessionID(), outlet.audio[0][new Random().nextInt(buffer_size)]);

		outlet.push();
	}

	private void updateKeysDown() {
		MidiInlet inlet = (MidiInlet)getInlet("frequency");
		if(inlet.available()) {
			MidiOutlet.MidiFrame midiFrame;
			for (Map.Entry<Integer, MidiOutlet.MidiFrame> entry : inlet.outlet.midi.entrySet()) {
				for (short[] midiPacket : entry.getValue()) {
					midiState.handle(midiPacket);
				}
			}
		}
	}

	@Override
	protected void onPrepare() {
		if(resampling == null && lnode.resamplingFactory != null && getOutlet("output") != null) {
			AudioOutlet outlet = (AudioOutlet)getOutlet("output");
			resampling = lnode.resamplingFactory.create(outlet.audio[0]);
		}

		if(resampling != null)
			resampling.rewind();
	}

	@Override
	protected void onDestroy() {}

	@Override
	public boolean isKeepingSessionAlive() {
		return active_tangent != null;
	}
}

File: src/main/kotlin/net/merayen/elastic/backend/architectures/local/nodes/from_midi_1/LProcessor.kt
package net.merayen.elastic.backend.architectures.local.nodes.from_midi_1

import net.merayen.elastic.backend.architectures.local.LocalProcessor
import net.merayen.elastic.backend.architectures.local.lets.MidiInlet
import net.merayen.elastic.backend.architectures.local.lets.SignalOutlet
import net.merayen.elastic.backend.architectures.local.nodes.poly_1.SessionKeeper
import net.merayen.elastic.backend.midi.MidiState
import net.merayen.elastic.backend.util.AudioUtil
import java.util.*

class LProcessor : LocalProcessor(), SessionKeeper {
	private var activeTangent: ShortArray? = null
	private val keysDown: ArrayList<ShortArray> = ArrayList()
	private var pos = Random().nextDouble() * Math.PI * 2
	private var volume = 0f

	private var done = false

	private val midiState: MidiState = object : MidiState() {
		override fun onPitchBendSensitivityChange(semitones: Float) {}

		override fun onKeyDown(tangent: Short, velocity: Float) {
			if (sustain == 0f) keysDown.add(currentMidiPacket) // TODO verify the sustain-comparison
			activeTangent = currentMidiPacket
		}

		override fun onKeyUp(tangent: Short) {
			keysDown.removeIf { it[1] == tangent }
			// TODO verify the sustain-comparison
			if (sustain == 0f)
				activeTangent = if (keysDown.isEmpty()) null else keysDown[keysDown.size - 1]
		}

		override fun onSustain(value: Float) {
			/*if (sustain && getCurrentMidiPacket()[2] == 0) {
				if (keys_down.isEmpty())
					active_tangent = null;
				else
					active_tangent = keys_down.get(keys_down.size() - 1);
			}

			sustain = getCurrentMidiPacket()[2] != 0;*/
		}

		override fun onVolumeChange(volume: Float) {}

		override fun onPitchChange(semitones: Float) {
			//System.out.println(getSessionID() + "\t" + semitones);
		}
	}

	override fun onInit() {}

	override fun onPrepare() {
		done = false
	}

	override fun onProcess() {
		if (done || !available())
			return

		val inlet = getInlet("in")

		if (inlet is MidiInlet) {
			var lastPosition = 0
			for ((midiPosition, midiFrame) in inlet.outlet.midi) {
				renderOutSignal(lastPosition, midiPosition)

				// Handle all the midi data on this exact time
				for (midiPacket in midiFrame)
					midiState.handle(midiPacket)

				lastPosition = midiPosition
			}

			// Render any remaining packets
			renderOutSignal(lastPosition, buffer_size)
		}

		getOutlet("frequency")?.push()
		getOutlet("amplitude")?.push()
		getOutlet("velocity")?.push()

		done = true
	}

	override fun onDestroy() {}

	override fun isKeepingSessionAlive() = activeTangent != null

	private fun renderOutSignal(from: Int, to: Int) {
		val frequency = (getOutlet("frequency") as? SignalOutlet)?.signal
		val amplitude = (getOutlet("amplitude") as? SignalOutlet)?.signal

		val activeTangent = activeTangent

		if (activeTangent != null) {
			// TODO Interpolate between midiframes too?
			if (frequency != null) {
				val freq = AudioUtil.midiNoteToFreq(activeTangent[1] + midiState.pitch).toFloat()
				for (i in from until to)
					frequency[i] = freq
			}

			if (amplitude != null) {
				val velocity = midiState.velocity
				val volume = midiState.volume
				for (i in from until to)
					amplitude[i] = velocity * volume
			}
		} else { // No keys down. Output silence
			if (frequency != null)
				for (i in from until to)
					frequency[i] = 0f

			if (amplitude != null)
				for (i in from until to)
					amplitude[i] = 0f
		}
	}
}